cmd/podman/images/scp.go

```diff
@@ -0,0 +1,383 @@
+	"golang.org/x/crypto/ssh"
+	"golang.org/x/crypto/ssh/agent"
+)
+
+var (
+	validScpFormats = []string{define.OCIManifestDir, define.OCIArchive, define.V2s2ManifestDir, define.V2s2Archive}
+)
+
+var (
+	saveScpDescription = `Securely copy an image from one host to another.`
+
+	imageScpCommand = &cobra.Command{
+		Use:   "scp [options] IMAGE [IMAGE...]",
+		Long:  saveScpDescription,
+		Short: "securely copy images",
+		RunE:  scp,
+		Args: func(cmd *cobra.Command, args []string) error {
+			if len(args) < 1 {
+				return errors.Errorf("need at least 1 argument")
+			}
+			format, err := cmd.Flags().GetString("format")
+			if err != nil {
+				return err
+			}
+			if !util.StringInSlice(format, validScpFormats) {
+				return errors.Errorf("format value must be one of %s", strings.Join(validScpFormats, " "))
+			}
+			return nil
+		},
+		ValidArgsFunction: common.AutocompleteImages,
+		Example:           `podman image scp myimage:latest otherhost::`,
+	}
+)
+
+var (
+	scpOpts entities.ImageScpOptions
+)
+
+func init() {
+	registry.Commands = append(registry.Commands, registry.CliCommand{
+		Command: imageScpCommand,
+		Parent:  imageCmd,
+	})
+	scpFlags(imageScpCommand)
+}
+
+func scpFlags(cmd *cobra.Command) {
+	flags := cmd.Flags()
+	formatFlagName := "format"
+	flags.StringVar(&scpOpts.Save.Format, formatFlagName, define.V2s2Archive, "Save image to oci-archive, docker-archive, docker-dir")
+	_ = cmd.RegisterFlagCompletionFunc(formatFlagName, common.AutocompleteImageSaveFormat)
+	flags.BoolVarP(&scpOpts.Save.Quiet, "quiet", "q", false, "Suppress the output")
+}
+
+func scp(cmd *cobra.Command, args []string) (finalErr error) {
+	var (
+		tags []string
+		err  error
+	)
+	if scpOpts.Save.Quiet { // set quiet for both load and save
+		scpOpts.Load.Quiet = true
+	}
+	scpOpts.ToFrom = true                 // default to loading to the ssh client
+	f, err := ioutil.TempFile("", "temp") // open temp dir for load/save output
+	if err != nil {
+		return err
+	}
+	scpOpts.Save.Output = f.Name()
+	scpOpts.Load.Input = scpOpts.Save.Output
+	if err := parse.ValidateFileName(saveOpts.Output); err != nil {
+		return err
+	}
+	confR, err := config.NewConfig("") // create a hand made config for the remote engine since we might use remote and native at once
+	if err != nil {
+		return errors.Wrapf(err, "could not make config")
+	}
+	abiEng, err := registry.NewImageEngine(cmd, args) // abi native engine
+	if err != nil {
+		return errors.Wrapf(err, "could not make abi engine")
+	}
+	url := ""
+	iden := ""
+	serv := map[string]config.Destination{}
+	cfg, err := config.ReadCustomConfig() // get ready to set ssh destination if necessary
+	if err != nil {
+		return err
+	}
+	conn := ""
+	if strings.Contains(args[0], "::") || len(args) == 2 { // if we have a connection name specified
+		if len(args) == 2 && !(strings.Contains(args[1], "::") && strings.Contains(args[0], "::")) { // if an image is specified, this mean we are loading from our client
```

Comment:
so @mtrmac you're suggesting I add more elements in `entities.ImageScpOptions`? That could allow us to just parse the args once.

Reply:
Isn’t `entities` a subpackage for the remote API, i.e. something that doesn’t quite apply to this code? (I don’t know _that_ much about how Podman code is organized.)

Maybe this could just be local variables within this function. OTOH collecting this all in `ImageScpOptions` might allow reducing the number of local variables.

---

The most eye-catching is the `len(args) == 2 && (!strings.Contains(…))` condition showing up in at least 3 instances.

Looking top-down, I was kind of expecting something like (intentionally using nonsense semantics):
```go
func argContainsSmiley(s string) (strings.Contains(s, ":-)")}

switch { // Figure out what the arguments mean
case len(args) == 1 && argContainsSmiley(args[0]): // this means … 
     smiley := parseSmiley(args[0])
     src := smiley
     dest := destForSmiley(smiley)
     return copyFromSmiley(src, dest)
case len(args) == 1 && !argContainsSmiley(args[0]): this means …
     text := parseText(args[0])
     src := text 
     dest := smileyForText(text)
     return copyToSmiley(src, dest)
case len(args) == 2 && argContainsSmiley(args[0]) && !argContainsSmiley(args[1]):
     src := parseSmiley(args[0])
     dest := parseText(args[1])
     return copyFromSmiley(src, dest)
…
default: // fail with internal error: unhandled case
}
```
with the various `copy…` functions sharing code via helper functions, or
```go
src := parseSource(args[0])
dest := nil
if len(args) == 2 { dest = parseDest(args[1]))
switch { // Figure out what the arguments mean
case len(args) == 1 && src.containsSmiley: // this means … 
     return copyFromSmiley(src, dest)
case len(args) == 1 && !src.ContainsSmiley: // this means …
     return copyToSmiley(src, dest)
case len(args) == 2 && src.containsSmiley && !dest.containsSmiley:  // this means …
     return copyFromSmiley(src, dest)
…
default: // fail with internal error: unhandled case
}
```
(not necessarily going _all the way_ to `src`/`dest` being full objects, `parse*` could return multiple values)

or maybe (closer to the current code?)
```go
// declare a scpMode enum

mode := scpModeUndefined
src, dest := nil
switch { // Figure out what the arguments mean
case len(args) == 1 && argContainsSmiley(args[0]): // this means … 
     smiley := parseSmiley(args[0])
     src = smiley
     dest = destForSmiley(smiley)
     mode = scpModeCopyFromSmiley
case len(args) == 1 && !argContainsSmiley(args[0]): // this means …
     text := parseText(args[0])
     src = text 
     dest = smileyForText(text)
     mode = scpModeCopyToSmiley
case len(args) == 2 && argContainsSmiley(args[0]) && !argContainsSmiley(args[1]): // this means …
     src = parseSmiley(args[0])
     dest = parseText(args[1])
     mode = scpModeCopyFromSmiley
…
}
// (All of the above could be a separate function returning (src, dest, mode), if it turned to be easier to read)
…
// Shared code for remote ssh setup, or whatever
…
switch mode {
case scpModeCopyFromSmiley:
    // Actually do the copy, don’t touch args again
…
default: // fail with internal error: unhandled case
}
```

The main point is that “determine what `mode` we are in” should only happen once, not all over the function; (although `mode` might be an explicit variable or implied by the control flow).